cpp/src/stream_compaction/distinct_count.cu

```diff
@@ -127,40 +129,39 @@ cudf::size_type distinct_count(table_view const& keys,
                                null_equality nulls_equal,
                                rmm::cuda_stream_view stream)
 {
-  auto const num_rows  = keys.num_rows();
+  auto const num_rows = keys.num_rows();
+  if (num_rows == 0) { return 0; }  // early exit for empty input
   auto const has_nulls = nullate::DYNAMIC{cudf::has_nested_nulls(keys)};
 
-  hash_map_type key_map{compute_hash_table_size(num_rows),
-                        cuco::empty_key{COMPACTION_EMPTY_KEY_SENTINEL},
-                        cuco::empty_value{COMPACTION_EMPTY_VALUE_SENTINEL},
-                        detail::hash_table_allocator_type{default_allocator<char>{}, stream},
-                        stream.value()};
-
   auto const preprocessed_input =
     cudf::experimental::row::hash::preprocessed_table::create(keys, stream);
-
   auto const row_hasher = cudf::experimental::row::hash::row_hasher(preprocessed_input);
   auto const hash_key   = experimental::compaction_hash(row_hasher.device_hasher(has_nulls));
-
-  auto const row_comp = cudf::experimental::row::equality::self_comparator(preprocessed_input);
-
-  auto iter = cudf::detail::make_counting_transform_iterator(
-    0, [] __device__(size_type i) { return cuco::make_pair(i, i); });
+  auto const row_comp   = cudf::experimental::row::equality::self_comparator(preprocessed_input);
 
   auto const comparator_helper = [&](auto const row_equal) {
+    using hasher_type = decltype(hash_key);
+    auto key_set      = cuco::experimental::static_set{
+      cuco::experimental::extent{static_cast<cudf::size_type>(compute_hash_table_size(num_rows))},
+      cuco::empty_key<cudf::size_type>{COMPACTION_EMPTY_KEY_SENTINEL},
+      row_equal,
```

Comment:
So the `static_set` is now constructed with a comparator? Is this a requirement?

Reply:
> So the static_set is now constructed with a comparator?

Yes, we are trying to align with `std::unordered_set` when possible. Similar to allocator and hasher, equality should be passed only once during construction (see [reference](https://en.cppreference.com/w/cpp/container/unordered_set/unordered_set)).

Reply:
Sorry to steal the show. Here is my suggestion to overcome the aforementioned issue: Combine both comparators into a thin proxy wrapper:
```
template<typename SelfComparator, typename TwoTableComparator>
struct comparator {
  SelfComparator self_comp;
  TwoTableComparator two_table_comp;

  bool operator()(lhs_index_type i, lhs_index_type j) const { return self_comp(i, j); } 
  bool operator()(lhs_index_type i, rhs_index_type j) const { return two_table_comp(i, j); } 
  bool operator()(rhs_index_type i, lhs_index_type j) const { return two_table_comp(i, j); } 
};
```

Reply:
When inserting, we should insert all keys of type `lhs_index_type` while when checking contains we will check keys of type `rhs_index_type` against the  `lhs_index_type` keys in the set. 

Theoretically, my proposal should work but  that depends on the detail implementation of `static_set`. If there is compile error with it then we may also need additional modification to `static_set`.

Reply:
> Theoretically, my proposal should work but that depends on the detail implementation of static_set.

Good point. cuco has an additional indirection layer to facilitate sentinel comparison thus the wrapper may not work as it is. To be tested.